db/db_impl_compaction_flush.cc

```diff
@@ -1115,6 +1073,59 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
   return s;
 }
 
+Status DBImpl::WaitUntilFlushWouldNotStallWrites(ColumnFamilyData* cfd) {
+  {
+    InstrumentedMutexLock l(&mutex_);
+    uint64_t orig_active_memtable_id = cfd->mem()->GetID();
+    WriteStallCondition write_stall_condition = WriteStallCondition::kNormal;
+    do {
+      if (write_stall_condition != WriteStallCondition::kNormal) {
+        TEST_SYNC_POINT("DBImpl::CompactRange:StallWait");
+        ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                       "[%s] CompactRange waiting on stall conditions to clear",
+                       cfd->GetName().c_str());
+        bg_cv_.Wait();
+      }
+      if (cfd->IsDropped() || shutting_down_.load(std::memory_order_acquire)) {
+        return Status::ShutdownInProgress();
+      }
+
+      uint64_t earliest_memtable_id =
+          std::min(cfd->mem()->GetID(), cfd->imm()->GetEarliestMemTableID());
+      if (earliest_memtable_id > orig_active_memtable_id) {
+        // We waited so long that the memtable we were originally waiting on was
+        // flushed.
+        break;
```

Comment:
If we reach this line, it means the original memtable we were originally waiting on was flushed. Then `WaitUntilFlushWouldNotStallWrites` return `Status::OK()` to the caller, i.e. `FlushMemTable`. It seems that `FlushMemTable` can return directly in this case. In the original code, we use `flush_needed` to avoid calling `FlushMemTable`.

Reply:
good catch :)

Reply:
I 'm thinking to just add `bool& flush_needed` param to the function to avoid that piece info in  addition to Status:: object